lib/runtime/transaction_validity.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package runtime
+
+import (
+	"errors"
+
+	"github.com/ChainSafe/gossamer/lib/transaction"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+// TransactionValidityError Information on a transaction's validity and, if valid,
+// on how it relates to other transactions. It is a result of the form:
+// Result<ValidTransaction, TransactionValidityError>
+type TransactionValidityError scale.VaryingDataType
+
+var (
+	errInvalidType     = errors.New("invalid validity type")
+	errInvalidResult   = errors.New("invalid error value")
+	errInvalidTypeCast = errors.New("invalid type cast")
+)
+
+// Set will set a VaryingDataTypeValue using the underlying VaryingDataType
+func (tve *TransactionValidityError) Set(val scale.VaryingDataTypeValue) (err error) {
+	vdt := scale.VaryingDataType(*tve)
+	err = vdt.Set(val)
+	if err != nil {
+		return
+	}
+	*tve = TransactionValidityError(vdt)
+	return
+}
+
+// Value will return value from underying VaryingDataType
+func (tve *TransactionValidityError) Value() (val scale.VaryingDataTypeValue) {
+	vdt := scale.VaryingDataType(*tve)
+	return vdt.Value()
+}
+
+func (tve *TransactionValidityError) Error() string {
+	invalidTxn, ok := tve.Value().(InvalidTransaction)
+	if !ok {
+		unknownTxn, ok2 := tve.Value().(UnknownTransaction)
+		if !ok2 {
+			return errInvalidTypeCast.Error()
+		}
+		return unknownTxn.Error()
+	}
+	return invalidTxn.Error()
+}
+
+// NewTransactionValidityError is constructor for TransactionValidityError
+func NewTransactionValidityError() TransactionValidityError {
+	vdt, err := scale.NewVaryingDataType(NewInvalidTransaction(), NewUnknownTransaction())
+	if err != nil {
+		panic(err)
+	}
+	return TransactionValidityError(vdt)
+}
+
+// UnmarshalTransactionValidity Takes the result of the validateTransaction runtime call and unmarshalls it
+// TODO use custom result issue #2780
+func UnmarshalTransactionValidity(res []byte) (*transaction.Validity, error) {
+	validTxn := transaction.Validity{}
+	txnValidityErrResult := NewTransactionValidityError()
+	txnValidityResult := scale.NewResult(validTxn, txnValidityErrResult)
+	err := scale.Unmarshal(res, &txnValidityResult)
+	if err != nil {
+		return nil, err
+	}
+	txnValidityRes, err := txnValidityResult.Unwrap()
+	if err != nil {
+		switch errType := err.(type) {
+		case scale.WrappedErr:
+			txnValidityErr, ok := errType.Err.(TransactionValidityError)
+			if !ok {
+				return nil, errInvalidTypeCast
+			}
+			return nil, &txnValidityErr
+		default:
+			return nil, errInvalidResult
+		}
+	} else {
+		switch validity := txnValidityRes.(type) {
+		case transaction.Validity:
+			return &validity, nil
+		default:
+			return nil, errInvalidType
+		}
+	}
```

Comment:
nit no need for the else since we return in all code paths in the if condition before.

Also let's dumb this type switch down since there is only one type checked.

And wrap the sentinel error to prevent callers to use `err == errInvalidType`, and force them to use `errors.Is(err, errInvalidType)`

```suggestion
	validity, ok := txnValidityRes.(transaction.Validity) {
	if ok {
		return &validity, nil
	}
	return nil, fmt.Errorf("%w", errInvalidType)
```